File: 01_GOF23/DP_examples/Python/01_Creational/Factory/Creator.py
from abc import ABC, abstractmethod
from Product import UnknownShape, Circle, Square, Rectangle

# 创建者接口
class ICreator(ABC):
    @abstractmethod
    def get_shape(self):
        pass

class UnknownShapeCreator(ICreator):
    def get_shape(self):
        return UnknownShape()

# 圆形创建者
class CircleCreator(ICreator):
    def get_shape(self):
        return Circle()

# 正方形创建者
class SquareCreator(ICreator):
    def get_shape(self):
        return Square()

# 矩形创建者
class RectangleCreator(ICreator):
    def get_shape(self):
        return Rectangle()

# 大圆形创建者
class BigCircleCreator(CircleCreator):
    def _set_radius(self, circle):
        print("Set Radius to 10")
        return circle
    
    def get_shape(self):
        return self._set_radius(super().get_shape())

File: 01_GOF23/DP_examples/Rust/01_Creational/Factory/creator.rs
use super::product::{IShape, ShapeKind, Circle, Square, Rectangle, UnknownShape};

pub trait ICreator {
    fn get_shape(&self) -> Box<dyn IShape>;
}

pub struct CircleCreator;

impl CircleCreator {
    pub const INSTANCE: CircleCreator = CircleCreator;
}
impl ICreator for CircleCreator {
    fn get_shape(&self) -> Box<dyn IShape> {
        Box::new(Circle)
    }
}

pub struct SquareCreator;

impl SquareCreator {
    pub const INSTANCE: SquareCreator = SquareCreator;
}
impl ICreator for SquareCreator {
    fn get_shape(&self) -> Box<dyn IShape> {
        Box::new(Square)
    }
}

pub struct RectangleCreator;

impl RectangleCreator {
    pub const INSTANCE: RectangleCreator = RectangleCreator;
}
impl ICreator for RectangleCreator {
    fn get_shape(&self) -> Box<dyn IShape> {
        Box::new(Rectangle)
    }
}

pub struct UnknownShapeCreator;

impl UnknownShapeCreator {
    pub const INSTANCE: UnknownShapeCreator = UnknownShapeCreator;
}
impl ICreator for UnknownShapeCreator {
    fn get_shape(&self) -> Box<dyn IShape> {
        Box::new(UnknownShape)
    }
}

pub struct BigCircleCreator;

impl BigCircleCreator {
    pub const INSTANCE: BigCircleCreator = BigCircleCreator;
    
    fn set_radius(&self) {
        println!("Set radius to 10");
    }
}
impl ICreator for BigCircleCreator {
    fn get_shape(&self) -> Box<dyn IShape> {
        self.set_radius();
        return CircleCreator::INSTANCE.get_shape()
    }
}